### PR TITLE
Introduce Kafka::asyncPublish() that will not flush on each send/batchSend but only once when the application is terminating

### PR DESCRIPTION
**Pull request by sash**
See discussion https://github.com/mateusjunges/laravel-kafka/discussions/309

### CODE REVIEW

File: src/Factory.php [hunk 1]
```diff
@@ -31,6 +31,19 @@ public function publish(string $broker = null): MessageProducer
         );
     }
 
+    /** Creates a new ProducerBuilder instance, setting brokers and topic. The producer will be flushed only when the application terminates, and doing SEND does not mean that the message was flushed! */
+    public function asyncPublish(string $broker = null): MessageProducer
```
Comment (sash) [inline src/Factory.php:35]: Alternatively a second argument could be introduced to the publish method: bool $async = false with the same effect but w/out the duplicated code in the factory...

Comment (mateusjunges) [reply src/Factory.php:35]: I prefer to have a dedicated method for this

File: src/Producers/Builder.php [hunk 2]
```diff
@@ -200,9 +204,14 @@ public function build(): Producer
             callbacks: $this->callbacks,
         );
 
-        return app(Producer::class, [
+        $res = app(Producer::class, [
```
Comment (mateusjunges) [inline src/Producers/Builder.php:207]:
```suggestion
        $producer = app(Producer::class, [
```

File: src/Producers/Builder.php [hunk 3]
```diff
@@ -200,9 +204,14 @@ public function build(): Producer
...
             'config' => $conf,
             'serializer' => $this->serializer,
+            'async' => $this->asyncProducer,
         ]);
+        if ($this->asyncProducer) {
+            $this->producer = $res;
```
Comment (mateusjunges) [inline src/Producers/Builder.php:213]:
```suggestion
            $this->producer = $producer;
```

File: src/Producers/Builder.php [hunk 4]
```diff
@@ -200,9 +204,14 @@ public function build(): Producer
...
+        }
+        return $res;
```
Comment (mateusjunges) [inline src/Producers/Builder.php:215]:
```suggestion
        return $producer;
```

File: src/Producers/Producer.php [hunk 5]
```diff
@@ -104,7 +114,9 @@ public function produceBatch(MessageBatch $messageBatch): int
             $produced++;
         }
 
-        $this->flush();
+        if (!$this->async) {
```
Comment (mateusjunges) [inline src/Producers/Producer.php:117]:
```suggestion
        if (! $this->async) {
```

File: src/Factory.php (hunk 1 above)
Comment (mateusjunges) [inline src/Factory.php:35]:
```suggestion
    public function asyncProducer(string $broker = null): MessageProducer
```

Currently, we use the `producer` method. I think it's better to call this one `asyncProducer` and keep it consistent

File: src/Facades/Kafka.php [hunk 6]
```diff
@@ -8,6 +8,7 @@
 
 /**
  * @method static \Junges\Kafka\Contracts\MessageProducer publish(string $broker = null)
+ * @method static \Junges\Kafka\Contracts\MessageProducer asyncPublish(string $broker = null)
```
Comment (mateusjunges) [inline src/Facades/Kafka.php:11]:
```suggestion
 * @method static \Junges\Kafka\Contracts\MessageProducer asyncProducer(string $broker = null)
```

File: docs/producing-messages/1-producing-messages.md [hunk 7]
```diff
@@ -12,4 +12,13 @@ Kafka::publish('broker')->onTopic('topic-name')
...
 This method returns a `ProducerBuilder` instance, which contains a few methods to configure your kafka producer. 
-The following lines describes these methods.
\ No newline at end of file
+The following lines describes these methods.
+
+If you are going to produce a lot of messages to different topics, please use the `asyncPublish` method on the `Junges\Kafka\Facades\Kafka` class
+```php
+use Junges\Kafka\Facades\Kafka;
+
+Kafka::asyncPublish('broker')->onTopic('topic-name')
```
Comment (mateusjunges) [inline docs/producing-messages/1-producing-messages.md:21]:
```suggestion
Kafka::asyncProducer('broker')->onTopic('topic-name')
```

File: docs/producing-messages/1-producing-messages.md [hunk 8]
```diff
@@ -12,4 +12,13 @@ Kafka::publish('broker')->onTopic('topic-name')
 ```
 
 This method returns a `ProducerBuilder` instance, which contains a few methods to configure your kafka producer. 
-The following lines describes these methods.
\ No newline at end of file
+The following lines describes these methods.
+
+If you are going to produce a lot of messages to different topics, please use the `asyncPublish` method on the `Junges\Kafka\Facades\Kafka` class
```
Comment (mateusjunges) [inline docs/producing-messages/1-producing-messages.md:17]:
```suggestion
If you are going to produce a lot of messages to different topics, please use the `asyncProducer` method on the `Junges\Kafka\Facades\Kafka` class
```

**Review by mateusjunges — CHANGES_REQUESTED** [review]
Thanks @sash! I left some small tweaks. I'm still testing this before I make any decisions, but it makes me want to remove support for batch publishing messages.

File: src/Factory.php (hunk 1 above)
Comment (mateusjunges) [reply src/Factory.php:35]: hmm my mistake, the method is actually called `publish`

**Comment by mateusjunges** [comment]
Merging into the `async` branch, I'll take it from here. Thank you @sash!

**Comment by sash** [comment]
I thought of something I didn't do - I need to have a singleton in order to keep a singular instance of the builder, so that even if you call Kafka::asyncPublish multiple times, you will still get the same instance that will have the same publisher. The current implementation creates multiple publishers and has each of them being flushed when the app terminates, and its up to the user to reuse the builder in his/hers code

**Pull request merged**